airbyte-cdk/python/airbyte_cdk/sources/file_based/file_types/csv_parser.py

```diff
@@ -5,18 +5,61 @@
 import csv
 from typing import Any, Dict, Iterable
 
+from airbyte_cdk.sources.file_based.config.file_based_stream_config import FileBasedStreamConfig
 from airbyte_cdk.sources.file_based.file_based_stream_reader import AbstractFileBasedStreamReader
 from airbyte_cdk.sources.file_based.file_types.file_type_parser import FileTypeParser
 from airbyte_cdk.sources.file_based.remote_file import RemoteFile
 
+DIALECT_NAME = "_config_dialect"
+
 
 class CsvParser(FileTypeParser):
-    async def infer_schema(self, file: RemoteFile, stream_reader: AbstractFileBasedStreamReader) -> Dict[str, Any]:
-        with stream_reader.open_file(file) as fp:
-            reader = csv.DictReader(fp)
-            return {field.strip(): {"type": ["null", "string"]} for field in next(reader)}
+    async def infer_schema(
+        self, config: FileBasedStreamConfig, file: RemoteFile, stream_reader: AbstractFileBasedStreamReader
+    ) -> Dict[str, Any]:
+        config_format = config.format
+        if config_format:
+            dialect_name = config.name + DIALECT_NAME
+            csv.register_dialect(
+                dialect_name,
+                delimiter=config_format.delimiter,
+                quotechar=config_format.quote_char,
+                escapechar=config_format.escape_char,
+                doublequote=config_format.double_quote,
+            )
+            with stream_reader.open_file(file) as fp:
+                # todo: the existing InMemoryFilesSource.open_file() test source doesn't currently require an encoding, but actual
+                #  sources will likely require one. Rather than modify the interface now we can wait until the real use case
+                reader = csv.DictReader(fp, dialect=dialect_name)
+                schema = {field.strip(): {"type": ["null", "string"]} for field in next(reader)}
+                csv.unregister_dialect(dialect_name)
+                return schema
+        else:
+            with stream_reader.open_file(file) as fp:
+                reader = csv.DictReader(fp)
+                return {field.strip(): {"type": ["null", "string"]} for field in next(reader)}
 
-    def parse_records(self, file: RemoteFile, stream_reader: AbstractFileBasedStreamReader) -> Iterable[Dict[str, Any]]:
-        with stream_reader.open_file(file) as fp:
-            reader = csv.DictReader(fp)
-            yield from reader
+    def parse_records(
+        self, config: FileBasedStreamConfig, file: RemoteFile, stream_reader: AbstractFileBasedStreamReader
+    ) -> Iterable[Dict[str, Any]]:
+        config_format = config.format
+        if config_format:
+            # Formats are configured individually per-stream so a unique dialect should be registered for each stream.
+            # Wwe don't unregister the dialect because we are lazily parsing each csv file to generate records
```

Comment:
want to call this one out a bit. the way dialects work, its not something you pass to the reader, but rather you register it with the csv package. it's not great that we leave this registered during the lifecycle of the sync, but we can't unregister it because the yield keeps going beyond the lifecycle of this method and we don't know when it finishes here.

One alternative I explored was adding interface methods like pre/post processing for the parser, that would allow us to instantiate the dialect before the start and after the end of reading. If we think its worth expanding the interface to do that we can, but I've tested this and we do isolate dialects per-stream like we should as is.

Reply:
could we check if it's already register, and if it is, fail if the dialect is different that the one from the config?

Reply:
I'm wondering about a solution that involves giving each stream its own parser object(s; in case we don't require a single file type in the future). What I'm thinking it would look like is that after the creation of the stream config, we'd  create a copy of the parser, register the dialect, and pass that into the stream on init. The non-csv parsers may not have the dialect pattern, so we'd possibly have to implement that sort of thing on their behalf. 

Alternatively we could not instantiate the parsers when they're passed into the source, and init them with *args, **kwargs, which are obtained from the config and which are used to apply the dialect.

I'm not opposed to the way you're doing it here but am thinking of a situation where future-us need to add a method to the parser and forget that we need to apply the dialect.

Reply:
I see, yeah it doesn't feel ideal to have to put the apply/create dialect as the first step for each of the parser's actions especially if we need to create more of them. However, it does also feel a even more hands on to have to add a new `apply_format()` method to the parser interface. And in that case, we would still have to make sure to call that method to every place we get the parser (`stream.get_parser(stream.config.file_type)`). 

In my opinion, I still think this is the change that gives us the most flexibility if we want to refactor how parsers get instantiated or how the format gets applied since we're not committing to more methods. So my vote would be leave it as is and extract this out later if we find it too unwieldly

Reply:
True, an `apply_format` method isn't the nicest approach. I'm totally fine leaving this as-is at least for now.